src/istgt_lu_disk.c

```diff
@@ -5552,68 +5564,214 @@ istgt_lu_disk_scsi_reserve(ISTGT_LU_DISK *spec, CONN_Ptr conn, ISTGT_LU_CMD_Ptr
-		MTX_UNLOCK(&spec->rcommonq_mtx);
-		clear_rcomm_cmd(rcomm_cmd);
+		MTX_UNLOCK(&spec->rq_mtx);
 		return -1;
 	}
-	TAILQ_INSERT_TAIL(&spec->rcommon_sendq, rcomm_cmd, send_cmd_next);
-	pthread_cond_signal(&spec->rcommonq_cond);
-	MTX_LOCK(&spec->luworker_rmutex[cmd->luworkerindx]);
-	MTX_UNLOCK(&spec->rcommonq_mtx);
-	if (rcomm_cmd->status == 0) //assuming status won't be 0 after IO completion
-		pthread_cond_wait(&spec->luworker_rcond[cmd->luworkerindx], &spec->luworker_rmutex[cmd->luworkerindx]);
-	MTX_UNLOCK(&spec->luworker_rmutex[cmd->luworkerindx]);
-	MTX_LOCK(&spec->rcommonq_mtx);
-	MTX_LOCK(&rcomm_cmd->rcommand_mtx);
-	status = rcomm_cmd->status;
-	if(status == -1 )  {
-		ISTGT_LOG("STATUS = -1\n");
-		if(rcomm_cmd->completed) {
-			MTX_UNLOCK(&rcomm_cmd->rcommand_mtx);
-			clear_rcomm_cmd(rcomm_cmd);
-		} else {
-			rcomm_cmd->completed = true;
-			MTX_UNLOCK(&rcomm_cmd->rcommand_mtx);
+
+	build_rcomm_cmd;
+
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		if(replica == NULL) {
+			REPLICA_LOG("Replica not present");
+			MTX_UNLOCK(&spec->rq_mtx);
+			exit(EXIT_FAILURE);
 		}
-		cmd->data = NULL;
-		rc = -1;
-		MTX_UNLOCK(&spec->rcommonq_mtx);
-		return -1;
+
+		/*
+		 * If there are some healthy replica then send read command
+		 * to all healthy replica else send read command to all
+		 * degraded replica.
+		 */
+		if (spec->healthy_rcount &&
+		    rcomm_cmd->opcode == ZVOL_OPCODE_READ) {
+			/*
+			 * If there are some healthy replica then don't send
+			 * a read command to degraded replica
+			 */
+			if (replica->state == ZVOL_STATUS_DEGRADED)
+				continue;
+			else
+				cmd_sent = true;
+		}
+
+		rcomm_cmd->copies_sent++;
+		build_rcmd();
+		put_to_mempool(&replica->cmdq, rcmd);
+		eventfd_write(replica->data_eventfd, 1);
+
+		if (cmd_sent)
+			break;
 	}
-	rc = cmd->data_len = rcomm_cmd->data_len;
-	cmd->data = rcomm_cmd->data;
-	if(rcomm_cmd->completed) {
-		MTX_UNLOCK(&rcomm_cmd->rcommand_mtx);
-		MTX_UNLOCK(&spec->rcommonq_mtx);
-		clear_rcomm_cmd(rcomm_cmd);
-	} else {
-		rcomm_cmd->completed = true;
-		MTX_UNLOCK(&rcomm_cmd->rcommand_mtx);
-		MTX_UNLOCK(&spec->rcommonq_mtx);
+
+	TAILQ_INSERT_TAIL(&spec->rcommon_waitq, rcomm_cmd, wait_cmd_next);
+
+	MTX_UNLOCK(&spec->rq_mtx);
+
+	// now wait for command to complete
+	while (1) {
+		MTX_LOCK(&spec->luworker_rmutex[cmd->luworkerindx]);
+		// check for status of rcomm_cmd
+		if (check_for_command_completion(spec, rcomm_cmd, cmd)) {
+			if (rcomm_cmd->status == -1)
+				rc = -1;
+			else
+				rc = cmd->data_len = rcomm_cmd->data_len;
+			put_to_mempool(&spec->rcommon_deadlist, rcomm_cmd);
+
+			MTX_LOCK(&spec->rq_mtx);
+			TAILQ_REMOVE(&spec->rcommon_waitq, rcomm_cmd, wait_cmd_next);
+			MTX_UNLOCK(&spec->rq_mtx);
+
+			MTX_UNLOCK(&spec->luworker_rmutex[cmd->luworkerindx]);
+			break;
+		}
+		now = time(NULL);
+		abstime.tv_sec = now + 1;
+		abstime.tv_nsec = 0;
+
+		pthread_cond_timedwait(&spec->luworker_rcond[cmd->luworkerindx], &spec->luworker_rmutex[cmd->luworkerindx],
```

Comment:
I don't think we miss any signals now as lock is taken while signaling..
however, just add in comments the work that need to be done to remove locking here